File: 907-koko-eating-bananas/koko-eating-bananas.cpp
class Solution {
public:
    long long func(vector<int>& piles,int hour){
        long long total_hours=0;
        for(int i=0;i<piles.size();i++){
            total_hours+=ceil((double)piles[i]/hour);
        }
        return total_hours;
    }
    int minEatingSpeed(vector<int>& piles, int h) {
        int low=1;
        int high=*max_element(piles.begin(), piles.end());
        int ans=high;
        while(low<=high){
            int mid=low+(high-low)/2;
            long long total_hours=func(piles,mid);
            if(total_hours<=h){
                ans=mid;
                high=mid-1;
            }
            else{
                low=mid+1;
            }
        }
        return ans;
    }
};

File: 907-koko-eating-bananas/koko-eating-bananas.py
# class Solution:
#     def can_koko_eat(self,mini,piles,h):
#         count=0
#         for i in range(0,len(piles)):
#             count+=(math.ceil(piles[i]/mini))
#         return count<=h
#     def minEatingSpeed(self, piles: List[int], h: int) -> int:
#         high=max(piles)
#         for i in range(1,high+1):
#             if(self.can_koko_eat(i,piles,h)):
#                 return i

class Solution:
    def can_koko_eat(self,mini,piles,h):
        count=0
        for i in range(0,len(piles)):
            count+=(math.ceil(piles[i]/mini))
        return count<=h
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        low=1
        high=max(piles)
        while(low<=high):
            mid=(low+high)//2
            if(self.can_koko_eat(mid,piles,h)):
                res=mid
                high=mid-1
            else:
                low=mid+1
        return res
        